kbs/src/token/jwk.rs

```diff
@@ -0,0 +1,166 @@
+use serde_json::Value;
+use std::fs::File;
+use std::io::BufReader;
+use std::path::Path;
+use std::result::Result::Ok;
+use std::str::FromStr;
+use thiserror::Error;
+
+const OPENID_CONFIG_URL_SUFFIX: &str = ".well-known/openid-configuration";
+
+#[derive(Error, Debug)]
+pub enum JwksGetError {
+    #[error("Invalid source path: {0}")]
+    SourcePath(String),
+    #[error("Failed to access source: {0}")]
+    SourceAccess(String),
+    #[error("Failed to deserialize source data: {0}")]
+    SourceDeserializeJson(String),
+}
+
+#[derive(Deserialize)]
+struct OpenIDConfig {
+    jwks_uri: String,
+}
+
+pub struct JwkAttestationTokenVerifier {
+    trusted_certs: Option<jwk::JwkSet>,
+}
+
+pub async fn get_jwks_from_file_or_url(p: &str) -> Result<jwk::JwkSet, JwksGetError> {
+    match Url::parse(p) {
+        Err(e) => Err(JwksGetError::SourcePath(e.to_string())),
+        Ok(mut url) if url.scheme() == "https" => {
+            url.set_path(OPENID_CONFIG_URL_SUFFIX);
+
+            let oidc = get(url.as_str())
+                .await
+                .map_err(|e| JwksGetError::SourceAccess(e.to_string()))?
+                .json::<OpenIDConfig>()
+                .await
+                .map_err(|e| JwksGetError::SourceDeserializeJson(e.to_string()))?;
+
+            let jwkset = get(oidc.jwks_uri)
+                .await
+                .map_err(|e| JwksGetError::SourceAccess(e.to_string()))?
+                .json::<jwk::JwkSet>()
+                .await
+                .map_err(|e| JwksGetError::SourceDeserializeJson(e.to_string()))?;
+
+            Ok(jwkset)
+        }
+        Ok(url) if url.scheme() == "file" && Path::new(url.path()).exists() => {
+            let file = File::open(url.path())
+                .map_err(|e| JwksGetError::SourceAccess(format!("open {}: {}", url.path(), e)))?;
+            let reader = BufReader::new(file);
+
+            serde_json::from_reader(reader)
+                .map_err(|e| JwksGetError::SourceDeserializeJson(e.to_string()))
+        }
+        Ok(url) => Err(JwksGetError::SourcePath(format!(
+            "unsupported scheme {} (must be either file or https) or missing file path",
+            url.scheme()
+        ))),
+    }
+}
+
+impl JwkAttestationTokenVerifier {
+    pub async fn new(config: &AttestationTokenVerifierConfig) -> Result<Self> {
+        let trusted_certs = match &config.trusted_certs_paths {
+            Some(paths) => {
+                let mut keyset = jwk::JwkSet { keys: Vec::new() };
+
+                for path in paths.iter() {
+                    match get_jwks_from_file_or_url(path).await {
+                        Ok(mut jwkset) => keyset.keys.append(&mut jwkset.keys),
+                        Err(e) => log::warn!("error getting JWKS: {:?}", e),
+                    }
+                }
+
+                match keyset.keys.len() {
+                    0 => None,
+                    _ => Some(keyset),
+                }
+            }
+            None => None,
```

Comment:
I wouldn't use a match here since there is really no logic in this case. You can just use `map` instead.

Reply:
Agreed. With the type change of `trusted_certs_paths`, we can simply write as
```rust
    pub async fn new(config: AttestationTokenVerifierConfig) -> Result<Self> {
        let keys = stream::iter(config.trusted_certs_paths)
            .filter_map(|p| async move {
                match get_jwks_from_file_or_url(&p).await {
                    Ok(jwkset) => Some(jwkset.keys),
                    Err(e) => {
                        warn!("error getting JWKS: {e:?}");
                        None
                    }
                }
            })
            .concat()
            .await;

        Ok(Self { trusted_certs: jwk::JwkSet { keys } })
    }
```

where `futures_util` crate is needed.

Reply:
Right, I guess I blindly followed the `CoCo` approach so it ended up looking the same

Reply:
> where `futures_util` crate is needed.

the proposal looks nice but I wonder if this is justified for the functionality here?

Reply:
What the `stream::iter` does, is try to do `get_jwks_from_file_or_url` concurrently if multiple paths in `trusted_certs_paths` are given. This makes sense to me. wdyt?

Reply:
My use-case is with one URL only but I understand we may have several `trusted_certs_path` in the future where parallelism *might* help to improve the KBS startup time.